lib/core/viper/module.dart

```diff
@@ -0,0 +1,30 @@
+import 'package:flutter/cupertino.dart';
+import 'package:streams_provider/streams_provider.dart';
+import 'package:survey/services/locator/locator_service.dart';
+
+export 'package:survey/core/extensions/build_context.dart';
+export 'package:survey/core/extensions/stream.dart';
+part 'presenter.dart';
+part 'interactor.dart';
+part 'router.dart';
+part 'view.dart';
+
+abstract class Module<V extends View, I extends Interactor, P extends Presenter,
+    R extends Router> extends StatelessWidget {
+  late final V view;
+  late final I interactor = locator.get();
+  late final P presenter = locator.get();
+  late final R router = locator.get();
```

Comment:
I have a question, should `Module` need to know `Interactor` and `Router`? 🤔  Where will we call `interaction.action()`, directly inside `Module` or `Presenter`?

Reply:
> I have a question, should Module need to know Interactor and Router?

You mean remove these lines and init them inside `void assembly(V view)` ?
```dart
  late final I interactor = locator.get();
  late final R router = locator.get();
```

> Where will we call interaction.action(), directly inside Module or Presenter?

It should be called inside presenter only

Reply:
> You mean remove these lines and init them inside `void assembly(V view)` ?
> 
> ```dart
>   late final I interactor = locator.get();
>   late final R router = locator.get();
> ```

Yes, if that is the place to use `interactor` and `router`. And Module will become:
````dart
Module<V extends View, P extends Presenter>
```
Can we? Sorry I still don't fully understand your structure so I'm not sure 😅 

> > Where will we call interaction.action(), directly inside Module or Presenter?
> 
> It should be called inside presenter only

Hmm, if so, can we init them directly inside presenter?

Reply:
## Overview
![Paper Notes 6 2](https://user-images.githubusercontent.com/17875522/117974025-48cf3d80-b357-11eb-9f54-8bf0d24bffb7.png)

## Answer
Yea, we can. But I want a module is a thing that holds everything and passes the router, interactor references to the presenter. So that, router, interactor, and presenter will belong to the module.
![Paper Notes 8](https://user-images.githubusercontent.com/17875522/117976234-bda37700-b359-11eb-8cdc-0f4b9ed61684.png)

And If we create an interactor, router inside a presenter, it seems they belong to a presenter rather than a module.
![Paper Notes 9](https://user-images.githubusercontent.com/17875522/117976299-d14edd80-b359-11eb-8fe3-9b0b64fe680b.png)

Reply:
@markgravity thanks for the detail explanation 🙇  yeah, same same but different.